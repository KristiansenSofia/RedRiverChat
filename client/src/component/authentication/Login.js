import React, { Component } from 'react'
import { Link, Redirect } from 'react-router-dom'

// Import NPM-modules
import Typography from 'material-ui/Typography'
import TextField from 'material-ui/TextField'
import Button from 'material-ui/Button'
import axios from 'axios/index'
import HttpsRedirect from 'react-https-redirect'

// Import styles. loginStyles for all imported components with a style attribute and CSS-file for classNames and id.
import {loginStyles} from '../../styles/AuthStyles'
import '../../styles/Styles.css'

import {AzureServerUrl} from '../../utils/Config'
import {validateLogin} from '../../utils/FormValidation'

/**
 *  Login-component.
 *
 *  @author Jimmy
 */

class Login extends Component {
  constructor (props) {
    super(props)

    this.state = {
      userName: '',
      password: '',
      navigate: false
    }

    this.handleSubmit = this.handleSubmit.bind(this)
  }

    /**
     *  Handle form-input. Input are added to this.state.
     *
     *  @author Jimmy
     */

    handleChange = name => event => {
      this.setState({
        [name]: event.target.value
      })
    };

    /**
     *  Handle submit-button. A login-request is sent to server with form-input included.
     *
     *  @author Jimmy
     */

    handleSubmit () {
      let validation = validateLogin(this.state)
      if (validation !== false) {
        return this.props.openSnackBar(validation)
      } else {
        this.sendRequest()
          .then((response) => {
            localStorage.setItem('token', JSON.stringify(response.data.token))
            this.props.userLogin(response.data.token)
            this.setState({navigate: true})
            return this.props.openSnackBar('Välkommen ' + this.state.userName + '!')
          }).catch((err) => {
            if (err.response.status === 401) {
              return this.props.openSnackBar('Fel användarnamn eller lösenord!')
            }
            return this.props.openSnackBar('Något gick fel. Försök igen!')
          })
      }
    }

    /**
     *  Send login request to server
     *
     *  @author Jimmy
     */

    sendRequest () {
      let tempObj = {
        username: this.state.userName,
        password: this.state.password,
      }

      return axios({
        method: 'post',
        url: AzureServerUrl + '/api/account/login',
        data: JSON.stringify(tempObj),
        headers: {'Content-Type': 'application/json'}
      })
    }

    render () {
      const { navigate } = this.state

      if (navigate) {
        return <Redirect to='/' push />
      }

      if (this.props.state.isSignedIn === true) {
        return <Redirect to='/' />
      }

      return (
        <HttpsRedirect>
          <div className='Login'>
            <Typography
              variant='headline'
              color='default'
              align='left'
              style={loginStyles.title}
            >
                      Logga in
            </Typography>
            <form style={loginStyles.container} noValidate autoComplete='off'>
              <TextField
                id='userName'
                label='Användarnamn'
                required
                style={loginStyles.textField}
                value={this.state.userName}
                onChange={this.handleChange('userName')}
                margin='normal'
              />
              <TextField
                id='password'
                label='Lösenord'
                required
                style={loginStyles.textField}
                type='password'
                autoComplete='current-password'
                onChange={this.handleChange('password')}
                margin='normal'
              />
              <div className='LoginButton'>
                <Button variant='raised' style={loginStyles.button} onClick={this.handleSubmit}>
                              Logga in
                </Button>
                <div style={loginStyles.loginLinkContainer}>
                  <div style={loginStyles.loginLinkDivLeft}>
                    <Link style={loginStyles.loginLink} to='/register'>Registrera ny användare</Link>
                  </div>
                  <div style={loginStyles.loginLinkDivRight}>
                    <Link style={loginStyles.loginLink} to='/password'>Glömt lösenord?</Link>
                  </div>
                </div>
              </div>
            </form>
          </div>
        </HttpsRedirect>

      )
    }
}

export default Login
