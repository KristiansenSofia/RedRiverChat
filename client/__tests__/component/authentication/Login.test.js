import React from 'react';
import { MemoryRouter } from 'react-router';
import Enzyme from 'enzyme';
import {mount, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Login from '../../../src/component/authentication/Login';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

Enzyme.configure({ adapter: new Adapter() });

test('Login renders without crashing', () => {
  const component = shallow(
    <MemoryRouter>
      <Login state = {{ isSignedIn: false }} />
    </MemoryRouter>
  );

  expect(component).toMatchSnapshot();
});

test('Login should navigate to user page when user is signed in', () => {
  const component = shallow(
    <Login state = {{ isSignedIn: true }} />
  );

  component.instance().render();

  expect(component).toMatchObject(/Redirect to="\/"/);
});


test('Login handleChange should set userName to Banan', () => {
  const component = shallow(
    <Login state = {{ isSignedIn: false }} />
  );

  let event = {
    target: {
      value: 'Banan'
    }
  }

  component.instance().handleChange('userName')(event);
  const userState = component.state('userName');

  expect(userState).toEqual('Banan');
});


test('Login handleChange should set password to Password123', () => {
  const component = shallow(
    <Login state = {{ isSignedIn: false }} />
  );

  let event = {
    target: {
      value: 'Password123'
    }
  }

  component.instance().handleChange('password')(event);
  const passwordState = component.state('password');

  expect(passwordState).toEqual('Password123');
});


test('Login handleSubmit should return "Formuläret ej korrekt ifyllt!" when username is missing', () => {
  const baseProps = {
    openSnackBar: jest.fn(),
    state: {
      isSignedIn: false,
    }
  };

  const component = shallow(
    <Login {...baseProps} />
  );

  component.setState({
    userName: '',
    password: 'Password123',
    email: 'banan@example.com'
  });

  component.instance().handleSubmit();

  expect(baseProps.openSnackBar).toHaveBeenCalledTimes(1);
});


test('Login handleSubmit should return "Formuläret ej korrekt ifyllt!" when password is missing', () => {
  const baseProps = {
    openSnackBar: jest.fn(),
    state: {
      isSignedIn: false,
    }
  };

  const component = shallow(
    <Login {...baseProps} />
  );

  component.setState({
    userName: 'Banan',
    password: '',
    email: 'banan@example.com'
  });

  component.instance().handleSubmit();

  expect(baseProps.openSnackBar).toHaveBeenCalledTimes(1);
});


test('Login handleSubmit should return "Formuläret ej korrekt ifyllt!" when email is missing', () => {
  const baseProps = {
    openSnackBar: jest.fn(),
    state: {
      isSignedIn: false,
    }
  };

  const component = shallow(
    <Login {...baseProps} />
  );

  component.setState({
    userName: 'Banan',
    password: 'Password123',
    email: ''
  });

  component.instance().handleSubmit();

  expect(baseProps.openSnackBar).toHaveBeenCalledTimes(1);
});


test('Login sendRequest should respond with status 200 when user exist', (done) => {
  let mock = new MockAdapter(axios);
  mock.onPost('https://redserver.azurewebsites.net/api/account/login').reply(200);

  const component = shallow(
    <Login state = {{ isSignedIn: false }} />
  );

  component.setState({
    userName: 'Banan',
    password: 'Password123',
    email: 'banan@example.com'
  });

  // Call sendRequest
  component.instance().sendRequest().then(response => {
    expect(response.status).toEqual(200);
    done();
  });
});


test('Login sendRequest should respond with status 401 when user does not exist', (done) => {
  let mock = new MockAdapter(axios);
  mock.onPost('https://redserver.azurewebsites.net/api/account/login').reply(401);

  const component = shallow(
    <Login state = {{ isSignedIn: false }} />
  );

  component.setState({
    userName: 'Sofia',
    password: 'Password123',
    email: 'sofia@example.com'
  });

  // Call sendRequest
  component.instance().sendRequest().catch(error => {
    expect(error.message).toEqual('Request failed with status code 401');
    expect(error.response.status).toEqual(401);
    done();
  });
});
