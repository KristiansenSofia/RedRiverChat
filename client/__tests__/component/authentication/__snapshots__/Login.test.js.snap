// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders without crashing 1`] = `
<div
  className="Login"
>
  <h1
    className="MuiTypography-root-1 MuiTypography-headline-6 MuiTypography-alignLeft-13"
    style={
      Object {
        "marginLeft": 30,
      }
    }
  >
    Logga in
  </h1>
  <form
    autoComplete="off"
    noValidate={true}
    style={
      Object {
        "display": "flex",
        "flexWrap": "wrap",
        "justifyContent": "space-around",
      }
    }
  >
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-25 MuiFormControl-marginNormal-26"
      onBlur={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "marginBottom": 20,
          "marginLeft": 30,
          "marginRight": 30,
          "width": 300,
        }
      }
    >
      <label
        className="MuiFormLabel-root-34 MuiInputLabel-root-29 MuiInputLabel-formControl-30 MuiInputLabel-animated-33"
        data-shrink={false}
        htmlFor="name"
      >
        Användarnamn
      </label>
      <div
        className="MuiInput-root-40 MuiInput-formControl-41 MuiInput-underline-44"
      >
        <input
          aria-invalid={false}
          aria-required={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-48"
          defaultValue={undefined}
          disabled={false}
          id="name"
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={undefined}
          rows={undefined}
          type="text"
          value={undefined}
        />
      </div>
    </div>
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-25 MuiFormControl-marginNormal-26"
      onBlur={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "marginBottom": 20,
          "marginLeft": 30,
          "marginRight": 30,
          "width": 300,
        }
      }
    >
      <label
        className="MuiFormLabel-root-34 MuiInputLabel-root-29 MuiInputLabel-formControl-30 MuiInputLabel-animated-33"
        data-shrink={false}
        htmlFor="password"
      >
        Lösenord
      </label>
      <div
        className="MuiInput-root-40 MuiInput-formControl-41 MuiInput-underline-44"
      >
        <input
          aria-invalid={false}
          aria-required={false}
          autoComplete="current-password"
          autoFocus={undefined}
          className="MuiInput-input-48 MuiInput-inputType-52"
          defaultValue={undefined}
          disabled={false}
          id="password"
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={undefined}
          rows={undefined}
          type="password"
          value={undefined}
        />
      </div>
    </div>
    <div
      className="LoginButton"
    >
      <button
        className="MuiButtonBase-root-69 MuiButton-root-54 MuiButton-raised-59"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "height": 60,
            "marginTop": 30,
            "width": 300,
          }
        }
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-55"
        >
          Logga in
        </span>
        <span
          className="MuiTouchRipple-root-72"
        />
      </button>
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
            "marginTop": 20,
            "width": 300,
          }
        }
      >
        <div
          style={
            Object {
              "height": 100,
              "textAlign": "left",
              "width": 150,
            }
          }
        >
          <a
            href="/register"
            onClick={[Function]}
            style={
              Object {
                "color": "#757575",
                "textDecoration": "none",
              }
            }
          >
            Registrera ny användare
          </a>
        </div>
        <div
          style={
            Object {
              "height": 100,
              "textAlign": "right",
              "width": 150,
            }
          }
        >
          <a
            href="/password"
            onClick={[Function]}
            style={
              Object {
                "color": "#757575",
                "textDecoration": "none",
              }
            }
          >
            Glömt lösenord?
          </a>
        </div>
      </div>
    </div>
  </form>
</div>
`;
